### PR TITLE
5140-Browse-senders-crash-on-characters

### LINKED ISSUE #5140: Browse senders crash on characters
If we try to browse the senders of a character it crashes.

`Instance of Character did not understand #first`

```
Character(Object)>>doesNotUnderstand: #first
ClyFullBrowser(ClyBrowserMorph)>>browseReferencesTo:from:
ClyFullBrowser(ClyBrowserMorph)>>browseReferencesTo:
ClyTextEditor>>sendersOf:
ClyTextEditor>>sendersOfIt
ClyTextEditor(RubSmalltalkEditor)>>sendersOfIt:
[ :target | target editor sendersOfIt: nil ] in RubSmalltalkEditor class>>buildShortcutsOn: in Block: [ :target | target editor sendersOfIt: nil ]
BlockClosure>>cull:
BlockClosure>>cull:cull:
BlockClosure>>cull:cull:cull:
KMCategoryBinding>>completeMatch:buffer:
[ :l | l completeMatch: self buffer: aBuffer ] in KMKeymap>>notifyCompleteMatchTo:buffer: in Block: [ :l | l completeMatch: self buffer: aBuffer ]
Array(SequenceableCollection)>>do:
KMKeymap>>notifyCompleteMatchTo:buffer:
KMKeymap>>onMatchWith:notify:andDo:
[ :entry | entry onMatchWith: anEventBuffer notify: aMatchListener andDo: aBlock ] in KMCategory>>onMatchWith:notify:andDo: in Block: [ :entry | entry onMatchWith: anEventBuffer notify...etc...
Set>>do:
KMCategory>>onMatchWith:notify:andDo:
KMCategoryBinding>>verifyMatchWith:notifying:thenDoing:
[ :aTarget | "nice hack to stop in the first listener" aTarget verifyMatchWith: anEventBuffer notifying: self thenDoing: [ ^ self ] ] in CmdKMDispatcher(KMDispatcher)>>dispatch: in Block: [ :aTarget | "nice hack to stop in the first liste...etc...
OrderedCollection>>do:
CmdKMDispatcher(KMDispatcher)>>dispatch:
CmdKMDispatcher>>dispatch:
KMTarget>>dispatch:
[ :targetToDispatch | 
targetToDispatch dispatch: KMBuffer uniqueInstance buffer copy.
aKeyboardEvent wasHandled ifTrue: [ ^ self ] ] in KMDispatchChain>>dispatch: in Block: [ :targetToDispatch | ...
KMDispatchChain>>do:
KMDispatchChain>>dispatch:
CmdKMDispatcher(KMDispatcher)>>dispatchKeystroke:
RubEditingArea(Morph)>>dispatchKeystrokeForEvent:
RubEditingArea(RubAbstractTextArea)>>handleKeystroke:

```

### PR DESCRIPTION
**Pull request by Ducasse**
Fixes: #5140 at least $n sender and implementors do not break anymore. Now ClyTextEditor should have rewritten a bit rubSmalltalkEditor to avoid duplication.

### CODE REVIEW

File: src/Calypso-SystemTools-QueryBrowser/ClyTextEditor.extension.st [hunk 1]
```diff
@@ -74,14 +75,13 @@ ClyTextEditor >> referencesToIt [
 
 { #category : #'*Calypso-SystemTools-QueryBrowser' }
 ClyTextEditor >> sendersOf: selectedSelector [
-	
+	selectedSelector isCharacter ifTrue: [ ^ self ].
```
Comment (dionisiydk) [inline src/Calypso-SystemTools-QueryBrowser/ClyTextEditor.extension.st]: Senders should normally work for any literals (try also cmd+n on integer like 10). We should just protect the check for class reference:
```Smalltalk
ClyBrowserMorph>>browseReferencesTo: aSymbol from: anEnvironment	
	...
	(aSymbol isSymbol and: [aSymbol first isUppercase]) ifTrue: [ 

```
So #isSymbol will solve the issue.

**Review by dionisiydk — CHANGES_REQUESTED** [review]

**Review by dionisiydk — APPROVED** [review]
thanks

**Comment by dionisiydk** [comment]
For the future we should improve involved low level methods:
```Smalltalk
CompiledCode>>hasLiteralThorough: literal
        .....	
        2 to: self numLiterals - 1 "exclude header and methodClass or outerCode"
	   do:[:index | | lit |
		(((lit := self objectAt: index) literalEqual: literal)
		 or: [lit isArray and: [lit hasLiteral: literal]]) ifTrue:
			[^ true].
		(lit isMemberOf: CompiledBlock) ifTrue: [ (lit hasLiteralThorough: literal) ifTrue: [ ^true ] ] ].
	^ false 
```
It looks quite bad. It should simply delegate to literal instead of all these type checks.
In that case VariableBinding as a literal could implement such message and we will avoid class name analysis by users of senders query. Actually it is not a senders query but a literal references query. 

I would also simplify names of those methods. I do not get this "Thorough" word. Simple #refersToLiteral: is much better. In fact it is existing method with deprecation comment inside. I think we should rethink it

**Comment by Ducasse** [comment]
100% with you. I created a separate issue because this code should be improved. #5154

**Comment by Ducasse** [comment]
https://github.com/pharo-project/pharo/issues/5154

**Pull request merged**